### ISSUE TITLE
fail2ban ufw script can't insert IPv6 bans

### ISSUE THREAD

**Issue by oerdnj** [fail2ban/fail2ban#2331]
### Environment:

_Fill out and check (`[x]`) the boxes which apply. If your Fail2Ban version is outdated, 
and you can't verify that the issue persists in the recent release, better seek support 
from the distribution you obtained Fail2Ban from_

- Fail2Ban version (including any possible distribution suffixes):
- OS, including release name/version: Debian stretch with backports
- [x] Fail2Ban installed via OS/distribution mechanisms: 0.10.2-2~bpo9+1
- [x] I checked 0.10.3.1 and 0.10.4 changelogs for possible fixes
- [x] You have not applied any additional foreign patches to the codebase
- [x] Some customizations were done to the configuration (provide details below is so)

### The issue:

When ufw is used as banaction, it ignores IPv6 bans because ufw can't insert IPv6 rules before IPv4 rules.  Therefore the rule is not inserted.

#### Steps to reproduce

Configure ufw as banaction and then manually ban any IPv6 address.

#### Expected behavior

IPv6 address/netblock inserted into ufw rules.

#### Observed behavior

* IPv6 address/netblock not inserted into ufw rules.

#### Any customizations done to /etc/fail2ban/ configuration
```
# cat /etc/fail2ban/jail.d/ufw.conf
[DEFAULT]
banaction = ufw
```

#### Workarounds

A custom script can be used to insert IPv6 addresses:
```
#!/bin/sh

# called as
# ufw-insert <insertpos> <blocktype> from <ip> to <destination> $app

position="$1"
blocktype="$2"
ip="$4"
destination="$6"
shift 6

if sipcalc "$ip" | head -1 | grep -q ipv6; then
    position="$(ufw status numbered | grep -F '22/tcp (v6)' | cut -f 1 -d ] | cut -f 2 -d [)"
fi

exec ufw insert "$position" "$blocktype" from "$ip" to "$destination" "$@"
```

**Comment by oerdnj** [comment]
When the custom script is used, the ufw output looks like this:
```
To                         Action      From
--                         ------      ----
Anywhere                   REJECT      171.224.0.0/11
Anywhere                   REJECT      14.224.0.0/11
Anywhere                   REJECT      118.68.192.0/19
Anywhere                   REJECT      116.108.0.0/14
Anywhere                   REJECT      103.47.132.0/22
22/tcp                     ALLOW       Anywhere
443/tcp                    ALLOW       Anywhere
80/tcp                     ALLOW       Anywhere
8060/tcp                   ALLOW       Anywhere
25/tcp                     ALLOW       Anywhere
993/tcp                    ALLOW       Anywhere
10050/tcp                  ALLOW       Anywhere
Anywhere (v6)              REJECT      2001:ee0::/32
Anywhere (v6)              REJECT      2402:800::/32
22/tcp (v6)                ALLOW       Anywhere (v6)
443/tcp (v6)               ALLOW       Anywhere (v6)
80/tcp (v6)                ALLOW       Anywhere (v6)
8060/tcp (v6)              ALLOW       Anywhere (v6)
25/tcp (v6)                ALLOW       Anywhere (v6)
993/tcp (v6)               ALLOW       Anywhere (v6)
10050/tcp (v6)             ALLOW       Anywhere (v6)

53                         ALLOW OUT   Anywhere
53 (v6)                    ALLOW OUT   Anywhere (v6)
```

(And yes, that's Vietnamese account spammers...)

**Comment by oerdnj** [comment]
That was my first thought as well, but that doesn't work if you have any existing `ALLOW` rules:

```
# ufw status numbered
Status: active

     To                         Action      From
     --                         ------      ----
[ 1] 22/tcp                     ALLOW IN    Anywhere
[ 2] DNS                        ALLOW IN    Anywhere
[ 3] Anywhere                   REJECT IN   192.0.2.1
[ 4] 22/tcp (v6)                ALLOW IN    Anywhere (v6)
[ 5] DNS (v6)                   ALLOW IN    Anywhere (v6)
[ 6] Anywhere (v6)              REJECT IN   2001:db8::1
```

and I checked the `iptables` rules directly and they are in wrong order:
```
Chain ufw-user-input (1 references)
 pkts bytes target     prot opt in     out     source               destination
    2   120 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:22
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:53 /* 'dapp_DNS' */
    0     0 ACCEPT     udp  --  *      *       0.0.0.0/0            0.0.0.0/0            udp dpt:53 /* 'dapp_DNS' */
    0     0 REJECT     all  --  *      *       192.0.2.1            0.0.0.0/0            reject-with icmp-port-unreachable
```

which makes the REJECT rule unusable for protocols with pre-existing hole in the firewall.

**Comment by sebres** [comment]
I know.
And this is pretty clear. Just IMHO the current way with `<insertpos>` is weird too, because targeting only one rule (as already said, the "bypass" rules may be different for both families) as well as would be wrong if someone inserts something before 1 manually and it should be definitely first rule.

Pity that ufw does not supports anchors resp. the grouping of rules in chains like iptables does it.

**Comment by oerdnj** [comment]
With sshguard, I had a custom `ip{,6}tables` named chain in ufw that was triggered before ufw inserted its rules and sshguard uses that particular chain.  Maybe it's something worth looking into for pre-ufw-0.36 installations.

Nevertheless, backporting ufw 0.36-1 to Debian stretch and mangling the rule to use `ufw prepend` works as expected.  Thank you for the tip.  Still a solution for people that want to use stock packages might be nice.

**Comment by oerdnj** [comment]
Here's improved shell snippet to get you a position of first IPv6 rule in the list:
```
ufw status numbered | grep -F ' (v6)' | head -1 | sed -e 's/\[\([0-9]*\)\].*/\1/'
```

**Comment by sebres** [comment]
Yes, basically it is possible, just what one should do it the insertpos's deviate by IPv4 and IPv6, for example there is a rule that forbid port 55 for IPv6 additionally (so basically no rule for IPv4):
```diff
  [ 1] 22/tcp                     ALLOW IN    Anywhere
  [ 2] DNS                        ALLOW IN    Anywhere
  [ 3] Anywhere                   REJECT IN   192.0.2.1
  [ 4] 22/tcp (v6)                ALLOW IN    Anywhere (v6)
+ [ 5] 55/tcp (v6)                 DENY IN    Anywhere (v6)
  [ 6] DNS (v6)                   ALLOW IN    Anywhere (v6)
  [ 7] Anywhere (v6)              REJECT IN   2001:db8::1
```
Then the calculated insert position for IPv6 would be "wrong".
Or just imagine someone added it after fail2ban is started (each time to reload of f2b-config hereafter is a bit ugly).
Therefore I don't know at the moment how it should be implemented better as using offset's.

**Comment by sebres** [comment]
hmm... may be #2337 or something similar could help to distinguish both sections (just for record).

**Comment by hmoffatt** [comment]
If I update my ufw to 0.36 to get the prepend command, what version of fail2ban do I need to make use of it?

**Comment by hmoffatt** [comment]
Ah I should just change actionban in action.d/ufw.conf

**Comment by blinkeye** [comment]
I also had a problem that a IPv6 attacker was not blocked. First check your current ufw version:

```bash
$ ufw version
```
If you have >= 0.36 you can either adapt `/etc/fail2ban/action.d/ufw.conf` or create a new file `/etc/fail2ban/action.d/ufw-custom.conf` with the necessary change:

```diff
$ diff ufw.conf ufw-custom.conf 
17c17
<             ufw insert <insertpos> <blocktype> from <ip> to <destination> $app
---
>             ufw prepend <blocktype> from <ip> to <destination> $app
23,26d22
< # Option: insertpos
< # Notes.:  The position number in the firewall list to insert the block rule
< insertpos = 1
< 
```
and use that in `/etc/fail2ban/jail.local`.

**Comment by thctlo** [comment]
https://bugs.debian.org/cgi-bin/bugreport.cgi?att=1;bug=909163;filename=0005-lp1368411.patch;msg=17 

fixed it for me. 
It shows in debian as fixed in 0.36-1, but it was not.

**Comment by ohader** [comment]
Just discovered the same issue having `ufw verson 0.35` and adjusted `/etc/fail2ban/action.d/ufw-ipv6.conf`

```
actionban = ipv6=$(echo '<ip>' | grep '^.*:.*')
            if [ "<ip>" = "$ipv6" ]; then grep=''; else grep='-v'; fi
            insertpos=$(ufw status numbered | grep '\[' | grep $grep '(v6)' | head -n 1 | cut -d ']' -f1 | cut -d '[' -f2);
            [ -n "<application>" ] && app="app <application>"
            ufw insert $insertpos <blocktype> from <ip> to <destination> $app
```

**Comment by gnosoz** [comment]
> Just discovered the same issue having `ufw verson 0.35` and adjusted `/etc/fail2ban/action.d/ufw-ipv6.conf`
> 
> ```
> actionban = ipv6=$(echo '<ip>' | grep '^.*:.*')
>             if [ "<ip>" = "$ipv6" ]; then grep=''; else grep='-v'; fi
>             insertpos=$(ufw status numbered | grep '\[' | grep $grep '(v6)' | head -n 1 | cut -d ']' -f1 | cut -d '[' -f2);
>             [ -n "<application>" ] && app="app <application>"
>             ufw insert $insertpos <blocktype> from <ip> to <destination> $app
> ```
> 
> (there are probably lots of optimizations to the `sh` script above)

It works like a charm and you deserve a medal, very nicely done and this should be integrated directly in future releases

**Comment by acsfer** [comment]
I can see errors too when banning IPv6 addresses.
```
ERROR   7fdf14c33170 -- exec: [ -n "" ] && app="app "
ufw insert 1 reject from 2a02:8428:8154:d101:6d79:59c0:d270:902d to any $app
ERROR   7fdf14c33170 -- stderr: "ERROR: Invalid position '1'"
ERROR   7fdf14c33170 -- returned 1
ERROR   Failed to execute ban jail 'nextcloud' action 'ufw' info 'ActionInfo({'ip': '2a02:8428:8154:d101:6d79:59c0:d270:902d', 'family': 'inet6', 'fid': <function Actions.ActionInfo.<lambda> at 0x7fdf14f93ca0>, 'raw-ticket': <function Actions.ActionInfo.<lambda> at 0x7fdf14f8d3a0>})': Error banning 2a02:8428:8154:d101:6d79:59c0:d270:902d
```

[This solution](https://github.com/fail2ban/fail2ban/issues/2331#issuecomment-511894383) works perfectly.

**Comment by sebres** [comment]
implemented in #3018
